src/main/scala/nl/knaw/dans/easy/ingest/EasyIngest.scala

```diff
@@ -163,15 +163,39 @@ object EasyIngest {
     if (dsSpec.dsLocation != "") {
       request = request.dsLocation(dsSpec.dsLocation)
     } else if (dsSpec.contentFile != "") {
-      request = sdo.listFiles.find(_.getName == dsSpec.contentFile) match {
-        case Some(file) => request.content(file)
-        case None => throw new RuntimeException(s"Couldn't find specified datastream: ${dsSpec.contentFile}")
+      request = (datastreamId, sdo.listFiles.find(_.getName == dsSpec.contentFile)) match {
+        case ("EMD", Some(file)) =>
+          // Note that this would change the ingested file's checksum, but it is only for the EMD datastream, which has no checksum
+          return executeAddRequestWithReplacement(request, file, "$sdo-id", pidDictionary(sdo.getName))
+        case (_, Some(file)) => request.content(file)
+        case (_, None) => throw new RuntimeException(s"Couldn't find specified datastream: ${dsSpec.contentFile}")
       }
     }
-
     request.execute().getLocation
   }
 
+  private def executeAddRequestWithReplacement(request: AddDatastream, file: File, placeholder: String, replacement:String): Try[URI] = {
+    val tmpFile = File.createTempFile(file.getName, null)
+    log.debug(s"Created temp file: '$tmpFile.getAbsolutePath'")
+
+    replacePlaceholderInFileCopy(file, tmpFile, placeholder, replacement)
+      .map(_ => request.content(tmpFile).execute().getLocation)
+      .eventually(() => {
+        log.debug(s"Deleting temp file: '$tmpFile.getAbsolutePath'")
+        FileUtils.deleteQuietly(tmpFile)
+      })
+  }
+
+  private def replacePlaceholderInFileCopy(src: File, dst: File, placeholder: String, replacement: String): Try[Unit] = Try {
+    // these files are assumed to be small enough to be read into memory without problems
+    val srcContent = FileUtils.readFileToString(src)
+    if (!srcContent.contains(placeholder))
+      throw new RuntimeException(s"Missing placeholder '$placeholder' in file: ${src.getAbsolutePath}")
```

Comment:
That would mean 

    run.doOnError(e => log.error(e.getMessage,e)).
      doOnSuccess(x => log.info(s"ingested: ${x.values.mkString(", ")}"))

in all our components with a run method. No doubt we could shorten that somehow too with some lazy evaluated toString for `x`.